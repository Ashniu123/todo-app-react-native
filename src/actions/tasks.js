import realm from 'realm';
import shortid from 'shortid';

import todoSchema from '../db/schema';

export const FETCH_DATA = 'FETCH_DATA';
export const ADD_ITEM = 'ADD_ITEM';
export const EDIT_ITEM = 'EDIT_ITEM';
export const REMOVE_ITEM = 'REMOVE_ITEM';
export const REMOVE_COMPLETED = 'REMOVE_ALL';
export const RESTORE_STATE = 'RESTORE_STATE';

export const fetchData = (data) => ({
  type: FETCH_DATA,
  data
});

export const startFetchData = () => {
  return (dispatch) => {
    realm
      .open({ schema: [todoSchema] })
      .then((realm) => {
        let fetchedData = realm.objects('Todolist'),
          data = [];
        if (fetchedData.length !== 0) {
          for (let item of fetchedData) {
            data.push({
              key: item.key,
              item: item.item,
              completed: item.completed
            });
          }
        }
        console.log(data);
        dispatch(fetchData(data));
        realm.close();
      })
      .catch((err) => {
        console.log(err);
      });
  };
};

export const addItem = (data) => ({
  type: ADD_ITEM,
  data
});

export const startAddItem = (item) => {
  return (dispatch) => {
    return new Promise((resolve, reject) => {
      realm
        .open({ schema: [todoSchema] })
        .then((realm) => {
          try {
            const data = {
              key: shortid.generate(),
              item,
              completed: false
            };
            realm.write(() => {
              realm.create('Todolist', data);
              console.log('created item');
              dispatch(addItem(data));
              resolve();
            });
          } catch (e) {
            reject('Error on create!');
          } finally {
            realm.close();
          }
        })
        .catch((err) => {
          console.log('Error creating!');
        });
    });
  };
};

export const editItem = (key) => ({
  type: EDIT_ITEM,
  key
});

export const startEditItem = (key) => {
  return (dispatch) => {
    realm
      .open({ schema: [todoSchema] })
      .then((realm) => {
        try {
          realm.write(() => {
            realm.create('Todolist', { key, completed: true }, true);
            console.log('updated item');
            dispatch(editItem(key));
          });
        } catch (e) {
          console.log('Error on edit!');
        } finally {
          realm.close();
        }
      })
      .catch((err) => {
        console.log('Error editing!');
      });
  };
};

export const removeItem = (key) => ({
  type: REMOVE_ITEM,
  key
});

export const startRemoveItem = (key) => {
  return (dispatch) => {
    realm
      .open({ schema: [todoSchema] })
      .then((realm) => {
        try {
          realm.write(() => {
            const itemToDelete = realm.objects('Todolist').filtered(`key = "${key}"`);
            realm.delete(itemToDelete);
            console.log('deleted item');
            dispatch(removeItem(key));
          });
        } catch (e) {
          console.log('Error on removing one!');
        } finally {
          realm.close();
        }
      })
      .catch((err) => {
        console.log('Error Removing!');
      });
  };
};

export const removeCompleted = () => ({
  type: REMOVE_COMPLETED
});

export const startRemoveCompleted = () => {
  return (dispatch) => {
    realm
      .open({ schema: [todoSchema] })
      .then((realm) => {
        try {
          realm.write(() => {
            const allCompletedItems = realm.objects('Todolist').filtered(`completed = true`);
            realm.delete(allCompletedItems);
            console.log('deleted all completed');
            dispatch(removeCompleted());
          });
        } catch (e) {
          console.log('Error on removing all!');
        } finally {
          realm.close();
        }
      })
      .catch((err) => {
        console.log('Error Removing all!');
      });
  };
};

export const restoreState = () => ({
  type: RESTORE_STATE
});

export const startRestoreState = () => {
  return (dispatch, getState) => {
    realm
      .open({ schema: [todoSchema] })
      .then((realm) => {
        try {
          const data = getState().tasks.prev;
          realm.write(() => {
            console.log('pastdata: ', data);
            realm.create('Todolist', data);
            console.log('restored item');
            dispatch(restoreState());
          });
        } catch (e) {
          console.log('Error on restoring state!');
        } finally {
          realm.close();
        }
      })
      .catch((err) => {
        console.log('Error Restoring state');
      });
  };
};
