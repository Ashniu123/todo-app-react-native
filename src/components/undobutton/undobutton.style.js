import { StyleSheet } from 'react-native';

export default StyleSheet.create({
  undobutton__view: {}
});
