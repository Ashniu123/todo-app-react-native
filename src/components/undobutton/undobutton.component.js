import React, { Component } from 'react';
import { Button } from 'react-native';
import { connect } from 'react-redux';
import * as Animatable from 'react-native-animatable';

import { startRestoreState } from '../../actions/tasks';

import styles from './undobutton.style';

export class UndoButton extends Component {
  constructor(props) {
    super(props);

    this.state = {
      show: false
    };

    this.handleOnPress = this.handleOnPress.bind(this);
    this.showTimeout = null;
  }

  componentDidUpdate(prevProps) {
    if (prevProps.data.length > this.props.data.length) {
      this.setState(
        {
          show: true
        },
        () => {
          this.showTimeout = null;
          this.showTimeout = setTimeout(() => {
            this.setState({
              show: false
            });
          }, 5000);
        }
      );
    }
  }

  handleOnPress() {
    this.setState(
      {
        show: false
      },
      this.props.startRestoreState
    );
  }

  render() {
    return (
      this.state.show && (
        <Animatable.View animation="fadeInUp" duration={500} style={styles.undobutton__view}>
          <Button
            color="#606060"
            title="UNDO"
            accessibilityLabel="Undo"
            onPress={this.handleOnPress}
          />
        </Animatable.View>
      )
    );
  }
}

const mapStateToProps = (state) => ({
  data: state.tasks.data
});

const mapDispatchToProps = (dispatch) => ({
  startRestoreState: () => dispatch(startRestoreState())
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(UndoButton);
