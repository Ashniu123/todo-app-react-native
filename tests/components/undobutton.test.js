import React from 'react';
import { shallow } from 'enzyme';

import { UndoButton } from '../../src/components/undobutton/undobutton.component';
import tasks from '../fixtures/tasks';

jest.useFakeTimers();

let wrapper, startRestoreState;

beforeEach(() => {
  startRestoreState = jest.fn();
  wrapper = shallow(<UndoButton startRestoreState={startRestoreState} data={tasks} />);
});

test('should render UndoButton component correctly', () => {
  expect(wrapper).toMatchSnapshot();
  wrapper.setState({ show: true });
  expect(wrapper).toMatchSnapshot();
});

test('should change state on delete item(s)', () => {
  expect(wrapper.state().show).toBe(false);
  wrapper.setProps({ data: tasks.splice(2) });
  expect(wrapper.state().show).toBe(true);
});

test('should show button on delete item(s) and disappear after 5 seconds', () => {
  expect(wrapper.state().show).toBe(false);
  wrapper.setProps({ data: tasks.splice(2) });
  expect(wrapper.state().show).toBe(true);
  jest.advanceTimersByTime(5000);
  expect(wrapper.state().show).toBe(false);
});

test('should restore state on button press', () => {
  wrapper.setState({ show: true });
  wrapper.find('Button').prop('onPress')();
  expect(startRestoreState).toHaveBeenCalled();
});
