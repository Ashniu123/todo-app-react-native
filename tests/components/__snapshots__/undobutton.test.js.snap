// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render UndoButton component correctly 1`] = `""`;

exports[`should render UndoButton component correctly 2`] = `
<withAnimatable(Component)
  animation="fadeInUp"
  delay={0}
  direction="normal"
  duration={500}
  iterationCount={1}
  iterationDelay={0}
  onAnimationBegin={[Function]}
  onAnimationEnd={[Function]}
  onTransitionBegin={[Function]}
  onTransitionEnd={[Function]}
  style={Object {}}
  useNativeDriver={false}
>
  <Button
    accessibilityLabel="Undo"
    color="#606060"
    onPress={[Function]}
    title="UNDO"
  />
</withAnimatable(Component)>
`;
